File: dist-sorting/java/Sorting.java
import java.util.*;
import java.util.stream.Collectors;

class SortingAnalyzer {
    private String algorithm;
    private List<Integer> original_arr;
    private List<Integer> arr;
    private long comparisons = 0;
    private long swaps = 0;

    private static final int MAX_VALUE = 1000;

    SortingAnalyzer(String algorithm, List<Integer> input) {
        this.algorithm = algorithm;
        this.original_arr = new ArrayList<>(input);
        this.arr = new ArrayList<>(input);
    }

    private void compare() {
        comparisons++;
    }

    private void swap(int i, int j) {
        Integer temp = arr.get(i);
        arr.set(i, arr.get(j));
        arr.set(j, temp);
        swaps++;
    }

    private void bubbleSort() {
        int n = arr.size();
        for (int i = 0; i < n; i++) {
            boolean swapped = false;
            for (int j = 0; j < n - i - 1; j++) {
                compare();
                if (arr.get(j) > arr.get(j + 1)) {
                    swap(j, j + 1);
                    swapped = true;
                }
            }
            if (!swapped) break;
        }
    }

    private void selectionSort() {
        int n = arr.size();
        for (int i = 0; i < n; i++) {
            int minIdx = i;
            for (int j = i + 1; j < n; j++) {
                compare();
                if (arr.get(j) < arr.get(minIdx)) {
                    minIdx = j;
                }
            }
            swap(i, minIdx);
        }
    }

    private void insertionSort() {
        for (int i = 1; i < arr.size(); i++) {
            int key = arr.get(i);
            int j = i - 1;
            while (j >= 0) {
                compare();
                if (arr.get(j) > key) {
                    arr.set(j + 1, arr.get(j));
                    swaps++;
                    j--;
                } else {
                    break;
                }
            }
            arr.set(j + 1, key);
        }
    }

    private void mergeSortHelper(int left, int right) {
        if (left >= right) return;
        int mid = left + (right - left) / 2;
        mergeSortHelper(left, mid);
        mergeSortHelper(mid + 1, right);
        merge(left, mid, right);
    }

    private void merge(int left, int mid, int right) {
        List<Integer> merged = new ArrayList<>();
        int i = left, j = mid + 1;
        while (i <= mid && j <= right) {
            compare();
            if (arr.get(i) <= arr.get(j)) {
                merged.add(arr.get(i++));
            } else {
                merged.add(arr.get(j++));
            }
            swaps++;
        }
        while (i <= mid) {
            merged.add(arr.get(i++));
            swaps++;
        }
        while (j <= right) {
            merged.add(arr.get(j++));
            swaps++;
        }
        for (int k = 0; k < merged.size(); k++) {
            arr.set(left + k, merged.get(k));
        }
    }

    private void mergeSort() {
        if (arr.size() > 0) {
            mergeSortHelper(0, arr.size() - 1);
        }
    }

    private int partition(int low, int high) {
        int pivot = arr.get(high);
        int i = low - 1;
        for (int j = low; j < high; j++) {
            compare();
            if (arr.get(j) < pivot) {
                i++;
                swap(i, j);
            }
        }
        swap(i + 1, high);
        return i + 1;
    }

    private void quickSortHelper(int low, int high) {
        if (low < high) {
            int pi = partition(low, high);
            quickSortHelper(low, pi - 1);
            quickSortHelper(pi + 1, high);
        }
    }

    private void quickSort() {
        if (arr.size() > 0) {
            quickSortHelper(0, arr.size() - 1);
        }
    }

    private void heapify(int i, int n) {
        while (true) {
            int largest = i;
            int left = 2 * i + 1;
            int right = 2 * i + 2;

            if (left < n) {
                compare();
                if (arr.get(left) > arr.get(largest)) {
                    largest = left;
                }
            }
            if (right < n) {
                compare();
                if (arr.get(right) > arr.get(largest)) {
                    largest = right;
                }
            }
            if (largest != i) {
                swap(i, largest);
                i = largest;
            } else {
                break;
            }
        }
    }

    private void heapSort() {
        int n = arr.size();
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(i, n);
        }
        for (int i = n - 1; i > 0; i--) {
            swap(0, i);
            heapify(0, i);
        }
    }

    private void countingSort() {
        if (arr.isEmpty()) return;
        int maxVal = arr.stream().mapToInt(Integer::intValue).max().orElse(0);
        int minVal = arr.stream().mapToInt(Integer::intValue).min().orElse(0);

        int[] count = new int[maxVal - minVal + 1];
        for (int v : arr) {
            count[v - minVal]++;
            compare();
        }

        int idx = 0;
        for (int i = 0; i < count.length; i++) {
            for (int j = 0; j < count[i]; j++) {
                arr.set(idx++, minVal + i);
                swaps++;
            }
        }
    }

    private void radixSort() {
        if (arr.isEmpty()) return;
        int maxVal = arr.stream().mapToInt(Integer::intValue).max().orElse(0);

        for (int exp = 1; maxVal / exp > 0; exp *= 10) {
            countingSortRadix(exp);
        }
    }

    private void countingSortRadix(int exp) {
        int[] count = new int[10];
        for (int v : arr) {
            count[(v / exp) % 10]++;
            compare();
        }

        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        List<Integer> output = new ArrayList<>(Collections.nCopies(arr.size(), 0));
        for (int i = arr.size() - 1; i >= 0; i--) {
            int idx = (arr.get(i) / exp) % 10;
            output.set(count[idx] - 1, arr.get(i));
            count[idx]--;
            swaps++;
        }
        arr = output;
    }

    private void bucketSort() {
        if (arr.isEmpty()) return;
        int maxVal = arr.stream().mapToInt(Integer::intValue).max().orElse(0);
        int minVal = arr.stream().mapToInt(Integer::intValue).min().orElse(0);

        int bucketCount = arr.size();
        List<List<Integer>> buckets = new ArrayList<>();
        for (int i = 0; i < bucketCount; i++) {
            buckets.add(new ArrayList<>());
        }

        for (int v : arr) {
            int idx;
            if (maxVal == minVal) {
                idx = 0;
            } else {
                idx = (int)((v - minVal) / (double)(maxVal - minVal) * (bucketCount - 1));
            }
            buckets.get(idx).add(v);
            compare();
        }

        int idx = 0;
        for (List<Integer> bucket : buckets) {
            insertionSortBucket(bucket);
            for (int v : bucket) {
                arr.set(idx++, v);
                swaps++;
            }
        }
    }

    private void insertionSortBucket(List<Integer> arr) {
        for (int i = 1; i < arr.size(); i++) {
            int key = arr.get(i);
            int j = i - 1;
            while (j >= 0 && arr.get(j) > key) {
                arr.set(j + 1, arr.get(j));
                j--;
            }
            arr.set(j + 1, key);
        }
    }

    private static final int MIN_RUN = 32;

    private void timInsertionSort(int left, int right) {
        for (int i = left + 1; i <= right; i++) {
            int key = arr.get(i);
            int j = i - 1;
            while (j >= left) {
                compare();
                if (arr.get(j) > key) {
                    arr.set(j + 1, arr.get(j));
                    swaps++;
                    j--;
                } else {
                    break;
                }
            }
            arr.set(j + 1, key);
        }
    }

    private void timMerge(int left, int mid, int right) {
        List<Integer> leftArr = new ArrayList<>(arr.subList(left, mid + 1));
        List<Integer> rightArr = new ArrayList<>(arr.subList(mid + 1, right + 1));
        int i = 0, j = 0, k = left;
        while (i < leftArr.size() && j < rightArr.size()) {
            compare();
            if (leftArr.get(i) <= rightArr.get(j)) {
                arr.set(k, leftArr.get(i++));
            } else {
                arr.set(k, rightArr.get(j++));
            }
            swaps++;
            k++;
        }
        while (i < leftArr.size()) {
            arr.set(k++, leftArr.get(i++));
            swaps++;
        }
        while (j < rightArr.size()) {
            arr.set(k++, rightArr.get(j++));
            swaps++;
        }
    }

    private void timSort() {
        int n = arr.size();
        for (int i = 0; i < n; i += MIN_RUN) {
            int right = Math.min(i + MIN_RUN - 1, n - 1);
            timInsertionSort(i, right);
        }
        for (int size = MIN_RUN; size < n; size *= 2) {
            for (int start = 0; start < n; start += size * 2) {
                int mid = start + size - 1;
                int right = Math.min(start + size * 2 - 1, n - 1);
                if (mid < right) {
                    timMerge(start, mid, right);
                }
            }
        }
    }

    private void introSortHelper(int low, int high, int maxDepth) {
        if (high - low <= 1) return;
        if (maxDepth == 0) {
            insertionSort();
            return;
        }

        int mid = low + (high - low) / 2;
        int pivot = arr.get(mid);
        List<Integer> left = new ArrayList<>(), middle = new ArrayList<>(), right = new ArrayList<>();

        for (int i = low; i < high; i++) {
            compare();
            if (arr.get(i) < pivot) {
                left.add(arr.get(i));
            } else if (arr.get(i) == pivot) {
                middle.add(arr.get(i));
            } else {
                right.add(arr.get(i));
            }
            swaps++;
        }

        int idx = low;
        for (int v : left) {
            arr.set(idx++, v);
        }
        for (int v : middle) {
            arr.set(idx++, v);
        }
        for (int v : right) {
            arr.set(idx++, v);
        }

        introSortHelper(low, low + left.size(), maxDepth - 1);
        introSortHelper(low + left.size() + middle.size(), high, maxDepth - 1);
    }

    private void introSort() {
        int maxDepth = 2 * (int)(Math.log(arr.size()) / Math.log(2));
        introSortHelper(0, arr.size(), maxDepth);
    }

    private void executeSort() {
        switch (algorithm) {
            case "bubble":
                bubbleSort();
                break;
            case "selection":
                selectionSort();
                break;
            case "insertion":
                insertionSort();
                break;
            case "merge":
                mergeSort();
                break;
            case "quick":
                quickSort();
                break;
            case "heap":
                heapSort();
                break;
            case "counting":
                countingSort();
                break;
            case "radix":
                radixSort();
                break;
            case "bucket":
                bucketSort();
                break;
            case "tim":
                timSort();
                break;
            case "intro":
                introSort();
                break;
        }
    }

    private boolean isSorted() {
        for (int i = 0; i < arr.size() - 1; i++) {
            if (arr.get(i) > arr.get(i + 1)) {
                return false;
            }
        }
        return true;
    }

    private String arrayToJsonString(List<Integer> list) {
        return "[" + list.stream().map(String::valueOf).collect(Collectors.joining(",")) + "]";
    }

    public String sort() {
        arr = new ArrayList<>(original_arr);
        comparisons = 0;
        swaps = 0;

        long start = System.nanoTime();
        executeSort();
        long end = System.nanoTime();

        double elapsed_ms = (end - start) / 1_000_000.0;

        StringBuilder sb = new StringBuilder();
        sb.append("{\"algorithm\":\"").append(algorithm).append("\",\"n\":").append(original_arr.size())
          .append(",\"input\":").append(arrayToJsonString(original_arr))
          .append(",\"output\":").append(arrayToJsonString(arr))
          .append(",\"actual_time_ms\":").append(String.format("%.6f", elapsed_ms))
          .append(",\"comparisons\":").append(comparisons)
          .append(",\"swaps\":").append(swaps)
          .append(",\"is_sorted\":").append(isSorted() ? "true" : "false")
          .append("}");

        return sb.toString();
    }
}

public class Sorting {
    private static final int MAX_ELEMENTS = 50;
    private static final int MAX_VALUE = 1000;

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: java Sorting <algorithm> [value1 value2 ...]");
            System.err.println("Algorithms: bubble, selection, insertion, merge, quick, heap, counting, radix, bucket, tim, intro");
            System.exit(1);
        }

        String algorithm = args[0];
        List<Integer> arr = new ArrayList<>();

        if (args.length > 1) {
            if (args.length - 1 > MAX_ELEMENTS) {
                System.err.println("Error: Maximum 50 elements allowed");
                System.exit(1);
            }
            for (int i = 1; i < args.length; i++) {
                try {
                    int val = Integer.parseInt(args[i]);
                    if (val < 0 || val > MAX_VALUE) {
                        System.err.println("Error: Element values must be between 0 and " + MAX_VALUE);
                        System.exit(1);
                    }
                    arr.add(val);
                } catch (NumberFormatException e) {
                    System.err.println("Error: Invalid number format");
                    System.exit(1);
                }
            }
        } else {
            Random rand = new Random(42);
            for (int i = 0; i < MAX_ELEMENTS; i++) {
                arr.add(rand.nextInt(MAX_VALUE + 1));
            }
        }

        SortingAnalyzer analyzer = new SortingAnalyzer(algorithm, arr);
        System.out.println(analyzer.sort());
    }
}

File: dist-sorting/python/sorting.py
#!/usr/bin/env python3
"""
Sorting Algorithm Performance Analyzer
Implements 11 sorting algorithms with performance metrics and high-precision timing.
"""

import json
import sys
import time
import random
from copy import deepcopy
from math import log2

MAX_ELEMENTS = 50
MAX_VALUE = 1000

class SortingAnalyzer:
    def __init__(self, algorithm, arr):
        self.algorithm = algorithm
        self.original_arr = arr.copy()
        self.arr = arr.copy()
        self.comparisons = 0
        self.swaps = 0
        self.n = len(arr)
    
    def swap(self, i, j):
        """Swap elements and track swaps"""
        self.arr[i], self.arr[j] = self.arr[j], self.arr[i]
        self.swaps += 1
    
    def compare(self):
        """Track comparisons"""
        self.comparisons += 1
    
    def _bubble_sort(self):
        """Bubble Sort - O(n²)"""
        n = len(self.arr)
        for i in range(n):
            swapped = False
            for j in range(0, n - i - 1):
                self.compare()
                if self.arr[j] > self.arr[j + 1]:
                    self.swap(j, j + 1)
                    swapped = True
            if not swapped:
                break
    
    def _selection_sort(self):
        """Selection Sort - O(n²)"""
        n = len(self.arr)
        for i in range(n):
            min_idx = i
            for j in range(i + 1, n):
                self.compare()
                if self.arr[j] < self.arr[min_idx]:
                    min_idx = j
            self.swap(i, min_idx)
    
    def _insertion_sort(self):
        """Insertion Sort - O(n²)"""
        n = len(self.arr)
        for i in range(1, n):
            key = self.arr[i]
            j = i - 1
            while j >= 0:
                self.compare()
                if self.arr[j] > key:
                    self.arr[j + 1] = self.arr[j]
                    self.swaps += 1
                    j -= 1
                else:
                    break
            self.arr[j + 1] = key
    
    def _merge_sort(self):
        """Merge Sort - O(n log n)"""
        self.arr = self._merge_sort_helper(self.arr)
    
    def _merge_sort_helper(self, arr):
        """Recursive merge sort helper"""
        if len(arr) <= 1:
            return arr
        
        mid = len(arr) // 2
        left = self._merge_sort_helper(arr[:mid])
        right = self._merge_sort_helper(arr[mid:])
        return self._merge(left, right)
    
    def _merge(self, left, right):
        """Merge two sorted arrays"""
        result = []
        i = j = 0
        while i < len(left) and j < len(right):
            self.compare()
            if left[i] <= right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1
            self.swaps += 1
        result.extend(left[i:])
        self.swaps += len(left) - i
        result.extend(right[j:])
        self.swaps += len(right) - j
        return result
    
    def _quick_sort(self):
        """Quick Sort - O(n log n) average"""
        self.arr = self._quick_sort_helper(self.arr, 0, len(self.arr) - 1)
    
    def _quick_sort_helper(self, arr, low, high):
        """Recursive quick sort helper"""
        if low < high:
            pi = self._partition(arr, low, high)
            arr = self._quick_sort_helper(arr, low, pi - 1)
            arr = self._quick_sort_helper(arr, pi + 1, high)
        return arr
    
    def _partition(self, arr, low, high):
        """Partition for quick sort"""
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            self.compare()
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
                self.swaps += 1
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        self.swaps += 1
        return i + 1
    
    def _heap_sort(self):
        """Heap Sort - O(n log n)"""
        n = len(self.arr)
        for i in range(n // 2 - 1, -1, -1):
            self._heapify(i, n)
        
        for i in range(n - 1, 0, -1):
            self.arr[0], self.arr[i] = self.arr[i], self.arr[0]
            self.swaps += 1
            self._heapify(0, i)
    
    def _heapify(self, i, n):
        """Heapify subtree rooted at index i"""
        largest = i
        left = 2 * i + 1
        right = 2 * i + 2
        
        if left < n:
            self.compare()
            if self.arr[left] > self.arr[largest]:
                largest = left
        
        if right < n:
            self.compare()
            if self.arr[right] > self.arr[largest]:
                largest = right
        
        if largest != i:
            self.arr[i], self.arr[largest] = self.arr[largest], self.arr[i]
            self.swaps += 1
            self._heapify(largest, n)
    
    def _counting_sort(self):
        """Counting Sort - O(n + k)"""
        if len(self.arr) == 0:
            return
        
        max_val = max(self.arr)
        min_val = min(self.arr)
        range_val = max_val - min_val + 1
        count = [0] * range_val
        
        for num in self.arr:
            count[num - min_val] += 1
            self.compare()
        
        idx = 0
        for i in range(range_val):
            while count[i] > 0:
                self.arr[idx] = i + min_val
                self.swaps += 1
                idx += 1
                count[i] -= 1
    
    def _radix_sort(self):
        """Radix Sort - O(nk)"""
        if len(self.arr) == 0:
            return
        
        max_val = max(self.arr)
        exp = 1
        
        while max_val // exp > 0:
            self._counting_sort_radix(exp)
            exp *= 10
    
    def _counting_sort_radix(self, exp):
        """Counting sort for radix sort"""
        n = len(self.arr)
        output = [0] * n
        count = [0] * 10
        
        for i in range(n):
            index = (self.arr[i] // exp) % 10
            count[index] += 1
            self.compare()
        
        for i in range(1, 10):
            count[i] += count[i - 1]
        
        for i in range(n - 1, -1, -1):
            index = (self.arr[i] // exp) % 10
            output[count[index] - 1] = self.arr[i]
            count[index] -= 1
            self.swaps += 1
        
        for i in range(n):
            self.arr[i] = output[i]
    
    def _bucket_sort(self):
        """Bucket Sort - O(n + k)"""
        if len(self.arr) == 0:
            return
        
        max_val = max(self.arr)
        min_val = min(self.arr)
        bucket_count = len(self.arr)
        bucket_range = (max_val - min_val) / bucket_count if bucket_count > 0 else 1
        
        buckets = [[] for _ in range(bucket_count)]
        
        for num in self.arr:
            if max_val == min_val:
                index = 0
            else:
                index = int((num - min_val) / bucket_range) if bucket_range > 0 else 0
            index = min(index, bucket_count - 1)
            buckets[index].append(num)
            self.compare()
        
        self.arr = []
        for bucket in buckets:
            bucket.sort()
            self.arr.extend(bucket)
            self.swaps += len(bucket)
    
    MIN_RUN = 32
    
    def _tim_insertion_sort(self, left, right):
        """Insertion sort for TimSort runs"""
        for i in range(left + 1, right + 1):
            key = self.arr[i]
            j = i - 1
            while j >= left:
                self.compare()
                if self.arr[j] > key:
                    self.arr[j + 1] = self.arr[j]
                    self.swaps += 1
                    j -= 1
                else:
                    break
            self.arr[j + 1] = key
    
    def _tim_merge(self, left, mid, right):
        """Merge for TimSort"""
        left_arr = self.arr[left:mid + 1]
        right_arr = self.arr[mid + 1:right + 1]
        i = j = 0
        k = left
        while i < len(left_arr) and j < len(right_arr):
            self.compare()
            if left_arr[i] <= right_arr[j]:
                self.arr[k] = left_arr[i]
                i += 1
            else:
                self.arr[k] = right_arr[j]
                j += 1
            self.swaps += 1
            k += 1
        while i < len(left_arr):
            self.arr[k] = left_arr[i]
            self.swaps += 1
            i += 1
            k += 1
        while j < len(right_arr):
            self.arr[k] = right_arr[j]
            self.swaps += 1
            j += 1
            k += 1
    
    def _tim_sort(self):
        """Tim Sort - O(n log n)"""
        n = len(self.arr)
        for i in range(0, n, self.MIN_RUN):
            right = min(i + self.MIN_RUN - 1, n - 1)
            self._tim_insertion_sort(i, right)
        
        size = self.MIN_RUN
        while size < n:
            for start in range(0, n, size * 2):
                mid = start + size - 1
                right = min(start + size * 2 - 1, n - 1)
                if mid < right:
                    self._tim_merge(start, mid, right)
            size *= 2
    
    def _intro_sort(self):
        """Intro Sort - O(n log n)"""
        def intro_sort_helper(arr, max_depth):
            if len(arr) <= 1:
                return arr
            if max_depth == 0:
                return sorted(arr)  # Switch to heap sort
            
            pivot = arr[len(arr) // 2]
            left = [x for x in arr if x < pivot]
            mid = [x for x in arr if x == pivot]
            right = [x for x in arr if x > pivot]
            
            for _ in arr:
                self.compare()
            
            for item in left + mid + right:
                self.swaps += 1
            
            return intro_sort_helper(left, max_depth - 1) + mid + intro_sort_helper(right, max_depth - 1)
        
        max_depth = int(log2(len(self.arr))) * 2 if len(self.arr) > 0 else 0
        self.arr = intro_sort_helper(self.arr, max_depth)
    
    def is_sorted(self):
        """Check if array is sorted"""
        for i in range(len(self.arr) - 1):
            if self.arr[i] > self.arr[i + 1]:
                return False
        return True
    
    def sort(self):
        """Execute the selected sorting algorithm"""
        start_time = time.perf_counter()
        self._execute_sort()
        total_time = (time.perf_counter() - start_time) * 1000
        
        return {
            "algorithm": self.algorithm,
            "n": self.n,
            "input": self.original_arr,
            "output": self.arr,
            "actual_time_ms": round(total_time, 6),
            "comparisons": self.comparisons,
            "swaps": self.swaps,
            "is_sorted": self.is_sorted()
        }
    
    def _execute_sort(self):
        """Execute the selected sorting algorithm (helper method)"""
        if self.algorithm == 'bubble':
            self._bubble_sort()
        elif self.algorithm == 'selection':
            self._selection_sort()
        elif self.algorithm == 'insertion':
            self._insertion_sort()
        elif self.algorithm == 'merge':
            self._merge_sort()
        elif self.algorithm == 'quick':
            self._quick_sort()
        elif self.algorithm == 'heap':
            self._heap_sort()
        elif self.algorithm == 'counting':
            self._counting_sort()
        elif self.algorithm == 'radix':
            self._radix_sort()
        elif self.algorithm == 'bucket':
            self._bucket_sort()
        elif self.algorithm == 'tim':
            self._tim_sort()
        elif self.algorithm == 'intro':
            self._intro_sort()


def validate_input(arr):
    """Validate array constraints"""
    if len(arr) > MAX_ELEMENTS:
        print(f"Error: Maximum {MAX_ELEMENTS} elements allowed")
        sys.exit(1)
    
    for val in arr:
        if val < 0 or val > MAX_VALUE:
            print(f"Error: Element values must be between 0 and {MAX_VALUE}")
            sys.exit(1)


def mode_1(algorithm, values):
    """Mode 1: Sort custom values"""
    arr = [int(v) for v in values]
    validate_input(arr)
    
    analyzer = SortingAnalyzer(algorithm, arr)
    result = analyzer.sort()
    
    if result is None:
        print(f"Unknown algorithm: {algorithm}")
        sys.exit(1)
    
    print(json.dumps(result))


def mode_2(algorithm):
    """Mode 2: Performance testing with 50 random elements"""
    arr = [random.randint(0, MAX_VALUE) for _ in range(MAX_ELEMENTS)]
    
    analyzer = SortingAnalyzer(algorithm, arr)
    result = analyzer.sort()
    
    if result is None:
        print(f"Unknown algorithm: {algorithm}")
        sys.exit(1)
    
    print(json.dumps(result))


def main():
    """Main entry point"""
    if len(sys.argv) < 2:
        print("Usage: python sorting.py <algorithm> [value1 value2 ...]")
        print("Algorithms: bubble, selection, insertion, merge, quick, heap, counting, radix, bucket, tim, intro")
        sys.exit(1)
    
    algorithm = sys.argv[1]
    
    if len(sys.argv) > 2:
        # Mode 1: Custom values
        mode_1(algorithm, sys.argv[2:])
    else:
        # Mode 2: Performance testing
        mode_2(algorithm)


if __name__ == "__main__":
    main()
